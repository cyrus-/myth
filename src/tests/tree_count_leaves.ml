let s = "
type bool =
  | True
  | False

type tree =
  | Leaf
  | Node of tree * bool * tree

type nat =
  | O
  | S of nat

let rec sum (n1:nat) (n2:nat) : nat =
  match n1 with
  | O -> n2
  | S (n1) -> S (sum n1 n2)
;;

let tree_count_leaves : tree -> nat |>
  { Leaf => 1
  | Node (Leaf, True, Leaf) => 2
  | Node (Node (Leaf, True, Leaf), True, Leaf) => 3
  | Node (Leaf, True, Node (Leaf, True, Leaf)) => 3
  | Node (Node (Node (Leaf, True, Leaf), True, Leaf), True, Leaf) => 4
  | Node (Node (Leaf, True, Leaf), True, Node (Leaf, True, Leaf)) => 4
  | Node (Node (Leaf, True, Leaf), True, Node (Node (Leaf, True, Leaf), True, Node (Leaf, True, Leaf))) => 6
  } = ?"
