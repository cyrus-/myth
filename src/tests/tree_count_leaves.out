(* Generated Thu Nov 13 00:35:54 2014 *)
let tree_count_leaves : tree -> nat =
  let rec f1 (t1:tree) : nat =
    match t1 with
      | Leaf -> S (O)
      | Node (t2, b1, t3) -> sum (f1 t2) (f1 t3)
  in
    f1
;;
